### ISSUE TITLE
پیش‌بینی باقی‌مانده عمر (RUL) با استفاده از LSTM

### ISSUE THREAD

**Issue by FARIDDD123** [FARIDDD123/Compressor-Dashboard#30]
در این وظیفه، هدف ما توسعه‌ی یک مدل **شبکه عصبی بازگشتی (LSTM)** برای پیش‌بینی **باقی‌مانده عمر مفید (RUL - Remaining Useful Life)** یک تجهیز صنعتی، بر اساس داده‌های سنسوری آن است. مدل از **داده‌های سری زمانی** استفاده کرده و مقدار RUL را برای هر نمونه تخمین می‌زند.  

---

#### **۱. آماده‌سازی داده‌ها**
- دریافت داده‌های حسگرها از پایگاه داده **MySQL** یا یک فایل **CSV**  
- پردازش داده‌ها شامل **پاک‌سازی، نرمال‌سازی و تقسیم‌بندی** داده‌ها  
- شکل‌دهی داده‌ها به قالب مناسب برای **LSTM** (اندازه ورودی: `(batch_size, sequence_length, num_features)`)  

#### **۲. طراحی مدل LSTM**
- استفاده از یک لایه **LSTM** با ۶۴ نورون  
- استفاده از یک لایه **Dense** با خروجی ۱ برای پیش‌بینی مقدار RUL  
- تنظیم تابع **فعال‌سازی خطی (Linear)** برای لایه خروجی  
- کامپایل مدل با استفاده از **Adam Optimizer** و تابع **MAE** به‌عنوان **تابع هزینه**  

#### **۳. آموزش مدل**
- استفاده از مجموعه داده **آموزشی (X_train, y_train)**  
- تنظیم **تعداد ۱۰۰ دوره (epochs)** برای آموزش مدل  
- ارزیابی مدل روی داده‌های **آزمایشی (X_test, y_test)**  

---

### **کد پیاده‌سازی مدل LSTM برای پیش‌بینی RUL**  

```python
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense
from tensorflow.keras.optimizers import Adam

# تعریف مدل LSTM
model = Sequential()
model.add(LSTM(64, input_shape=(60, 10)))  # 60 گام زمانی، 10 ویژگی
model.add(Dense(1, activation='linear'))

# کامپایل مدل
model.compile(loss='mae', optimizer=Adam())

# آموزش مدل
model.fit(X_train, y_train, epochs=100, batch_size=64, validation_data=(X_test, y_test))
```

---

### **۴. خروجی مورد انتظار**
پس از آموزش مدل، انتظار داریم مقدار **باقی‌مانده عمر مفید (RUL)** برای هر نمونه داده محاسبه شود. مدل باید قادر باشد با دریافت مقادیر حسگرها، روند تغییرات را تشخیص داده و میزان فرسایش تجهیز را تخمین بزند.

---

### **۵. بهبود و بهینه‌سازی**
- تنظیم **ابعاد ورودی** (افزایش یا کاهش `sequence_length` و `num_features`)  
- تغییر **تعداد لایه‌ها و نورون‌ها** در LSTM  
- استفاده از **Dropout** برای جلوگیری از **overfitting**  
- تست مدل با **مجموعه داده‌های مختلف** برای بررسی دقت و تعمیم‌پذیری  

**هدف نهایی:** ایجاد یک مدل دقیق و بهینه برای **پیش‌بینی RUL** جهت **افزایش کارایی و کاهش هزینه‌های تعمیر و نگهداری تجهیزات صنعتی**.
